proposals/0000-replacing-cabal-custom-build.md

```diff
@@ -0,0 +1,1775 @@
+    `${output}`).
+
+In particular, such a design fits the needs of the Haskell Language Server,
+which needs to be able to:
+
+  1. Query the package for all its pre-build rules.
+  2. Find out all the rules that have become stale and need to be re-run.
+  3. Execute individual rules.
+
+However, the textual description of rules presented above suffers from some
+limitations that would make migrating existing packages with `Custom` build-type
+difficult. In particular, one often wants to allow rules to depend on each other
+in a more dynamic manner.  
+For example, consider how one might want to query an external executable in
+order to determine the dependency structure; say by running
+[`ghc -M`](https://downloads.haskell.org/ghc/latest/docs/users_guide/separate_compilation.html#makefile-dependencies)
+on a root Haskell module in order to compute a build graph.
+
+### Motivation: a simplistic first design
+
+To explain the design we have arrived at for fine-grained pre-build rules, let
+us first consider what a first draft design, which accommodates both the design
+of HLS as well as the existing `Custom` setup scripts, might look like:
+
+```haskell
+type TentativeRules env = env -> IO [TentativeRule]
+data TentativeRule = TentativeRule
+  { dependencies :: [Dependency]
+  , results :: [Result]
+  , action :: [ResolvedLocation]
+               -- ^ locations of __dependencies__ (determined by the build system)
+           -> [ResolvedLocation]
+              -- ^ locations for __results__ (chosen by the build system)
+           -> IO ()
+  }
+```
+
+That is, rules are specified by a function that takes in an environment
+(which in practice consists of information known to `Cabal` after configuring,
+e.g. `LocalBuildInfo`, `ComponentLocalBuildInfo`).
+This function returns an `IO` action that computes a list of rules. Each rule
+declares its inputs and outputs, and from this information arises the dependency
+structure between rules.  
+To run a rule, the build system must determine specific locations for all
+these inputs and outputs, and pass them to the `action` in order to
+execute the rule. For example, the build system would search for `blah/foo/bar.y`
+in the source directories of the project, and pass an absolute path to the
+file it found to the `action`.
+
+### Proposed design of rules
+
+There are several shortcomings with the above simplistic design:
+
+  1. it does not fit well with the proposed IPC interface for hooks:
+     namely, the build tool should be able to query the separate hooks
+     executable to obtain all the hooks that a package with `Hooks` `build-type`
+     provides. Doing so with the design proposed above would require a mechanism for
+     serialising and deserialising the `IO` actions that execute the rules, which in
+     practice would mean providing a DSL for `IO` actions that can be serialised,
+
+  2. it lacks information that would allow us to determine when the rules need
```

Comment:
At this point I started to go "why are we not just copying Shake?". In particular, I think it's pretty well-established how to deal with this kind of thing (see e.g. "Build Systems a la carte").

I am sure the authors are aware of this, so it would be really helpful to have a comparison to the existing prior art in this area to make it a bit clearer where the design comes from. It's a little hard to tell in isolation!

Reply:
I agree that a justification for why we are deviating from prior art is lacking from the current proposal; we will add that in the next iteration.

I would like to know what precisely you mean by "copying Shake" here? To me, the relevant question is rather "why are we not just copying Ninja?". I say that because pre-build rules should be structurally simple enough that they can be ingested by other build systems, just like Shake can ingest a Makefile or a Ninja file, constructing a build graph out of it. If we start having arbitrary monadicity inside the structure of the rules, then I don't know how another build tool such as HLS would be able to interface with the hooks; we need some way of exposing the dependency graph of rules externally.

Reply:
Copying Ninja might also be good! It's just really hard to know how to assess this when it seems to be its own special snowflake at the moment. Or, say, adopting the terminology of "Build Systems a la Carte" might help.

I don't see why arbitrary monadicity is even necessarily a problem. Tools with dynamic dependencies like Shake do need to tell the build system which dependencies were "discovered" when running a rule, but I don't see why you can't do that gracefully over an IPC connection or whatever. 

I think the problem

Reply:
Perhaps the issue is more with the _scheduler_? You want to do an up-front schedule, which requires no dynamic dependencies. I do agree that Shake's "suspending" approach to making this work is going to be tricky over IPC.

But to go on about this: it would be _very_ helpful to read something like "this system is basically Shake but with no dynamic dependencies, which we did for reason X".